From 9cadb19ec8309e5fa4b79259d2aea9396fc761ca Mon Sep 17 00:00:00 2001
From: "kenan.liu" <kenan.liu@amlogic.com>
Date: Tue, 21 Dec 2021 10:05:28 +0800
Subject: [PATCH] libav: CB0 implement gstlibav with ffmpeg in Gstreamer [1/1]

PD#SWPL-57554

Problem:
default disable ffmpeg-demux for not affect other project

Solution:
ENABLE_FFMPEG_DEMUX=TRUE means using ffmpeg-demux

Verify:
AP222
---
 ext/libav/gstavcodecmap.c | 17 ++++---
 ext/libav/gstavdemux.c    | 96 +++++++++++++++++++++------------------
 2 files changed, 61 insertions(+), 52 deletions(-)

diff --git a/ext/libav/gstavcodecmap.c b/ext/libav/gstavcodecmap.c
index 56a8a03..f011c71 100644
--- a/ext/libav/gstavcodecmap.c
+++ b/ext/libav/gstavcodecmap.c
@@ -3522,14 +3522,16 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
   GstCaps *caps = NULL;
 
   if (!strcmp (format_name, "mpeg")) {
-    caps = gst_caps_new_simple ("video/mpeg",
-        "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    //caps = gst_caps_new_simple ("video/mpeg",
+    //    "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    caps = gst_caps_from_string ("video/x-cdxa");
   } else if (!strcmp (format_name, "mpegts")) {
     caps = gst_caps_new_simple ("video/mpegts",
         "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
   } else if (!strcmp (format_name, "rm")) {
-    caps = gst_caps_new_simple ("application/x-pn-realmedia",
-        "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    //caps = gst_caps_new_simple ("application/x-pn-realmedia",
+    //    "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
+    caps = gst_caps_new_empty_simple ("application/vnd.rn-realmedia");
   } else if (!strcmp (format_name, "asf")) {
     caps = gst_caps_new_empty_simple ("video/x-ms-asf");
   } else if (!strcmp (format_name, "avi")) {
@@ -3547,7 +3549,7 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
         "systemstream", G_TYPE_BOOLEAN, TRUE, NULL);
   } else if (!strcmp (format_name, "4xm")) {
     caps = gst_caps_new_empty_simple ("video/x-4xm");
-  } else if (!strcmp (format_name, "matroska")) {
+  } else if (!strcmp (format_name, "matroska_webm")) {
     caps = gst_caps_new_empty_simple ("video/x-matroska");
   } else if (!strcmp (format_name, "ivf")) {
     caps = gst_caps_new_empty_simple ("video/x-ivf");
@@ -3561,7 +3563,7 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
     caps = gst_caps_new_empty_simple ("audio/x-ttafile");
   } else if (!strcmp (format_name, "aiff")) {
     caps = gst_caps_new_empty_simple ("audio/x-aiff");
-  } else if (!strcmp (format_name, "mov_mp4_m4a_3gp_3g2")) {
+  } else if (!strcmp (format_name, "mov_mp4_m4a_3gp_3g2_mj2")) {
     caps =
         gst_caps_from_string
         ("application/x-3gp; video/quicktime; audio/x-m4a");
@@ -3583,7 +3585,8 @@ gst_ffmpeg_formatid_to_caps (const gchar * format_name)
   } else if (!strcmp (format_name, "gif")) {
     caps = gst_caps_from_string ("image/gif");
   } else if (!strcmp (format_name, "ogg")) {
-    caps = gst_caps_from_string ("application/ogg");
+    //caps = gst_caps_from_string ("application/ogg");
+    caps = gst_caps_from_string ("video/ogg");
   } else if (!strcmp (format_name, "mxf") || !strcmp (format_name, "mxf_d10")) {
     caps = gst_caps_from_string ("application/mxf");
   } else if (!strcmp (format_name, "gxf")) {
diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index fa5fd4e..7b27b29 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -1982,6 +1982,7 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
   };
 
   void *i = 0;
+  const char *env;
 
   GST_LOG ("Registering demuxers");
 
@@ -2071,51 +2072,56 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
         !strcmp (in_plugin->name, "ivf"))
       register_typefind_func = FALSE;
 
-    /* Set the rank of demuxers known to work to MARGINAL.
-     * Set demuxers for which we already have another implementation to NONE
-     * Set All others to NONE*/
-    if (!strcmp (in_plugin->name, "wsvqa") ||
-        !strcmp (in_plugin->name, "wsaud") ||
-        !strcmp (in_plugin->name, "wc3movie") ||
-        !strcmp (in_plugin->name, "voc") ||
-        !strcmp (in_plugin->name, "tta") ||
-        !strcmp (in_plugin->name, "sol") ||
-        !strcmp (in_plugin->name, "smk") ||
-        !strcmp (in_plugin->name, "vmd") ||
-        !strcmp (in_plugin->name, "film_cpk") ||
-        !strcmp (in_plugin->name, "ingenient") ||
-        !strcmp (in_plugin->name, "psxstr") ||
-        !strcmp (in_plugin->name, "nuv") ||
-        !strcmp (in_plugin->name, "nut") ||
-        !strcmp (in_plugin->name, "nsv") ||
-        !strcmp (in_plugin->name, "mxf") ||
-        !strcmp (in_plugin->name, "mmf") ||
-        !strcmp (in_plugin->name, "mm") ||
-        !strcmp (in_plugin->name, "ipmovie") ||
-        !strcmp (in_plugin->name, "ape") ||
-        !strcmp (in_plugin->name, "RoQ") ||
-        !strcmp (in_plugin->name, "idcin") ||
-        !strcmp (in_plugin->name, "gxf") ||
-        !strcmp (in_plugin->name, "ffm") ||
-        !strcmp (in_plugin->name, "ea") ||
-        !strcmp (in_plugin->name, "daud") ||
-        !strcmp (in_plugin->name, "avs") ||
-        !strcmp (in_plugin->name, "aiff") ||
-        !strcmp (in_plugin->name, "4xm") ||
-        !strcmp (in_plugin->name, "yuv4mpegpipe") ||
-        !strcmp (in_plugin->name, "pva") ||
-        !strcmp (in_plugin->name, "mpc") ||
-        !strcmp (in_plugin->name, "mpc8") ||
-        !strcmp (in_plugin->name, "ivf") ||
-        !strcmp (in_plugin->name, "brstm") ||
-        !strcmp (in_plugin->name, "bfstm") ||
-        !strcmp (in_plugin->name, "gif") ||
-        !strcmp (in_plugin->name, "dsf") || !strcmp (in_plugin->name, "iff"))
-      rank = GST_RANK_MARGINAL;
-    else {
-      GST_DEBUG ("ignoring %s", in_plugin->name);
-      rank = GST_RANK_NONE;
-      continue;
+    env = getenv("ENABLE_FFMPEG_DEMUX");
+    if ((env) && (!strcmp(env, "TRUE"))) {
+        rank = GST_RANK_PRIMARY;
+    } else {
+        /* Set the rank of demuxers known to work to MARGINAL.
+         * Set demuxers for which we already have another implementation to NONE
+         * Set All others to NONE*/
+        if (!strcmp (in_plugin->name, "wsvqa") ||
+            !strcmp (in_plugin->name, "wsaud") ||
+            !strcmp (in_plugin->name, "wc3movie") ||
+            !strcmp (in_plugin->name, "voc") ||
+            !strcmp (in_plugin->name, "tta") ||
+            !strcmp (in_plugin->name, "sol") ||
+            !strcmp (in_plugin->name, "smk") ||
+            !strcmp (in_plugin->name, "vmd") ||
+            !strcmp (in_plugin->name, "film_cpk") ||
+            !strcmp (in_plugin->name, "ingenient") ||
+            !strcmp (in_plugin->name, "psxstr") ||
+            !strcmp (in_plugin->name, "nuv") ||
+            !strcmp (in_plugin->name, "nut") ||
+            !strcmp (in_plugin->name, "nsv") ||
+            !strcmp (in_plugin->name, "mxf") ||
+            !strcmp (in_plugin->name, "mmf") ||
+            !strcmp (in_plugin->name, "mm") ||
+            !strcmp (in_plugin->name, "ipmovie") ||
+            !strcmp (in_plugin->name, "ape") ||
+            !strcmp (in_plugin->name, "RoQ") ||
+            !strcmp (in_plugin->name, "idcin") ||
+            !strcmp (in_plugin->name, "gxf") ||
+            !strcmp (in_plugin->name, "ffm") ||
+            !strcmp (in_plugin->name, "ea") ||
+            !strcmp (in_plugin->name, "daud") ||
+            !strcmp (in_plugin->name, "avs") ||
+            !strcmp (in_plugin->name, "aiff") ||
+            !strcmp (in_plugin->name, "4xm") ||
+            !strcmp (in_plugin->name, "yuv4mpegpipe") ||
+            !strcmp (in_plugin->name, "pva") ||
+            !strcmp (in_plugin->name, "mpc") ||
+            !strcmp (in_plugin->name, "mpc8") ||
+            !strcmp (in_plugin->name, "ivf") ||
+            !strcmp (in_plugin->name, "brstm") ||
+            !strcmp (in_plugin->name, "bfstm") ||
+            !strcmp (in_plugin->name, "gif") ||
+            !strcmp (in_plugin->name, "dsf") || !strcmp (in_plugin->name, "iff"))
+          rank = GST_RANK_MARGINAL;
+        else {
+          GST_DEBUG ("ignoring %s", in_plugin->name);
+          rank = GST_RANK_NONE;
+          continue;
+        }
     }
 
     /* construct the type */
-- 
2.32.0

